Create CacheGuardian To Handle Fiber Removal

linhong-intel opened the pull request:
## What changes were proposed in this pull request?

Use a `refCount` to track fiber usage number
Only add fiber to a pending queue in onRemoval
Create a CacheGuardian to handle pending removal fibers
CacheGuardian will Wait fiber release if refCount > 0
Changed some code to ensure Fiber is release after using


## How was this patch tested?

Unit test pass, add a new test case

LuciferYang commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -46,13 +47,32 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
+            fiberCache.dispose()
+            _cacheSize.addAndGet(-notification.getValue.size())
+          } else {
+            logWarning(s"Remove Cache ${notification.getKey} failed")
```
in this else branch, fiberCache not dispose , _cacheSize not decrement , but the notification already remove from the cache, Is there a memory leak ?

linhong-intel replied:
Thanks, good finding. But the memory leak is not only on this fiber, we need cleanup all the caches before exception. Or the cached memory will never be released.

linhong-intel replied:
updated, the fix is in FiberCacheManager.get()

chenghao-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -88,9 +108,16 @@ object FiberCacheManager extends Logging {
+    if (exceptionFlag) {
+      // Here we don't check the refCount since we will throw exception
+      cache.asMap().values().asScala.foreach(_.dispose())
+      throw new OapException("Can't remove in-used fiber within 3 seconds")
```
Shall we set the `exceptionFlag` back to false?

chenghao-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -88,9 +108,16 @@ object FiberCacheManager extends Logging {
+      cache.asMap().values().asScala.foreach(_.dispose())
+      throw new OapException("Can't remove in-used fiber within 3 seconds")
+    }
+    fiberCache.refCount.incrementAndGet()
```
Or can we use the function instead of call the refCount directly? `fiberCache.open()` or `fiberCache.occupy()` etc.

chenghao-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -88,9 +108,16 @@ object FiberCacheManager extends Logging {
+    val fiberCache = cache.get(fiber, cacheLoader(fiber, conf))
+    if (exceptionFlag) {
+      // Here we don't check the refCount since we will throw exception
+      cache.asMap().values().asScala.foreach(_.dispose())
```
as we will throw exception, which will fails the current task, but doens't mean we have to release all of the cache isn't? even if we release all of the cache, don't forget to reset the `_cacheSize`.

chenghao-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -46,13 +47,32 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
+        try {
+          // lock will be released in await and acquired after thread wake up.
+          if (fiberCache.nonUsed.await(3000, TimeUnit.MILLISECONDS)) {
+            fiberCache.dispose()
```
This is really a danger operation, as will cause the null pointer exception (segment fault) in offheap memory access, and executor will crash without any hint.

linhong-intel commented:
@chenghao-intel @carsonwang @yueawang 
Let's review this PR. Thanks.

yueawang commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -92,17 +126,21 @@ object FiberCacheManager extends Logging {
+    .weigher(weigher)
+    .build[Fiber, FiberCache]()
 
   def get(fiber: Fiber, conf: Configuration): FiberCache = {
```
Does this func need synchronized?

linhong-intel replied:
Guava cache is thread safe, there is a lock inside.
If we use synchronized here, we will lose the concurrency.
synchronized is same to change `concurrencyLevel(4)` to `concurrencyLevel(1)`

linhong-intel replied:
Yes, it's synchronized :)

yueawang commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -46,6 +46,39 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
+        val fiberCache = queue.take()
+        logDebug(s"Removing fiber $fiberCache ...")
+        // Block if fiber is in use.
+        while (!fiberCache.dispose(3000, TimeUnit.MILLISECONDS)) {
```
Maybe name like disposeWithTimer or disposeWait sound better??

linhong-intel replied:
sure

yueawang commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -46,6 +46,39 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
+        // Block if fiber is in use.
+        while (!fiberCache.dispose(3000, TimeUnit.MILLISECONDS)) {
+          // Check memory usage every 3s while we are waiting fiber release.
+          if (queue.asScala.map(_.size()).sum > maxMemory) {
```
I thought the memory size in CacheGuardian just works as a temp buffer to prevent easy overflow of cache memory. Correct me if wrong.
Does the total size of those buffers in queue really matter here?

linhong-intel replied:
This log is useful for debugging.
We just don't want to use memory more than we requested from spark.
If this happens, usually something is wrong, like fiber not release for a very long time, or off-heap memory is too small.
The worst case is, we use too much memory and yarn will kill the executor.

carsonwang commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -46,6 +46,39 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
+  private val DEFAULT_QUEUE_INITIAL_CAPACITY = 11
+  // Ordering.by() makes sure non-used FiberCache dequeue first, used by test suite.
+  private[filecache] val removalPendingQueue = new PriorityBlockingQueue[FiberCache](
+    DEFAULT_QUEUE_INITIAL_CAPACITY, Ordering.by(_.refCount))
```
When you release a FiberCache, its refCount is updated. But its position in the priority queue will not be updated. Next time you take an element from the queue you are not going to get the FiberCache with lowest refCount.

linhong-intel replied:
I checked the implementation of `PriorityBlockingQueue`
sort happens both in `offer()` and `take()`. So, `take()` will make sure the the poped entry is the lowest at current queue.

carsonwang replied:
Really? Getting the element with highest priority should only be O(1) and removing it should be O(logN). Check this discussion  https://stackoverflow.com/questions/1871253/updating-java-priorityqueue-when-its-elements-change-priority

linhong-intel replied:
Thanks @carsonwang , you are right. I see, the sort in remove is after pop.
Anyway, Let's change to a simple queue without priority and optimize this later as @chenghao-intel  said.

carsonwang commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -46,6 +46,39 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
+        val fiberCache = queue.take()
+        logDebug(s"Removing fiber $fiberCache ...")
+        // Block if fiber is in use.
+        while (!fiberCache.tryDispose(3000, TimeUnit.MILLISECONDS)) {
```
It seems we will block at here forever if this fiberCache can not be disposed? It is possible that other fiberCaches be added to the queue whose refCount is 0. So can we go through all the elements in the queue and only dispose the fiberCache whose refCount is 0?

linhong-intel replied:
Yes, this is what we talked at Monday. We have 2 options:
1. check pending fibers one by one and wait until refCount is 0
2. check pending fibers one by one and skip fibers with refCount > 0

We have chosen option#1, but option#2 is also acceptable.

carsonwang replied:
The problem of #1 is that we may wait for a long time or forever, and the thread does nothing. At the same time, more fibers (refCount can be 0) are going to be added to the removal queue and we have more chance to be OOM.

chenghao-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala:
```diff
@@ -33,9 +36,53 @@ import org.apache.spark.unsafe.types.UTF8String
 
 // TODO: make it an alias of MemoryBlock
 trait FiberCache {
```
```scala
trait FiberCache {
  private var _refCount = 0
  def synchronized occupy(): Unit = { _refCount += 1 }
  def synchronized release(): Unit = { 
    _refCount -= 1
    notifyAll()
  }

  def synchronized tryDispose(): Unit = {
    while(_refCount > 0) {
      wait(100) // 100 ms or even less
      log.warning(s"Fiber Cache Dispose waiting detected for ${this.toString()}")
    }
    realDispose()
  }

  private[this] def realDispose(): Unit = { release the memory }
}
```

Do you think if above code works?

linhong-intel replied:
of course it works.

linhong-intel replied:
As I know, synchronized function is locked the whole object. I mean, `occupy`, `release`, `tryDispose` can't happen at the same time.
Then, if we enter `tryDispose() -> while(_refCount > 0)`, it will loop forever, since `release` cannot be called.

linhong-intel replied:
In another way, if synchronized only lock it's own function, then `occupy` and `tryDispose` can happen at the same time, right after `while(_refCount > 0)` and before `realDispose()`. So we have to introduce another lock to make sure this won't happen.
Maybe it's `def get(fiber: Fiber, conf: Configuration): FiberCache = synchronized`

chenghao-intel replied:
yes, that's by design. and once `wait(100)` called, the `synchronized` lock will be released. see: https://docs.oracle.com/javase/tutorial/essential/concurrency/guardmeth.html

the `get()` and `put()` will be called in different threads.

linhong-intel replied:
Ah, I see. Ok, I'll change to this.

chenghao-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -46,6 +46,39 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
+      maxMemory: Long) extends Thread {
+    override def run(): Unit = {
+      // Loop forever
+      while (true) {
```
We probably need a way to release the CacheGuardian / FiberCacheManager just for ease of unit testing, but can leave this for the next PR or release.

linhong-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -92,17 +126,21 @@ object FiberCacheManager extends Logging {
+    val fiberCache = cache.get(fiber, cacheLoader(fiber, conf))
+    // Avoid loading a fiber larger than MAX_WEIGHT / 4, 4 is concurrency number
+    assert(fiberCache.size() <= MAX_WEIGHT * MB / 4, "Can't cache fiber larger than MAX_WEIGHT / 4")
+    fiberCache.occupy()
```
Yes, this is a problem.
How about:
```
def get(fiber: Fiber, conf: Configuration): FiberCache = synchronized {
}
```
Since `onRemoval` is called inside `get()`, the operation is sequential.

chenghao-intel replied:
so probably we need to add the `synchronized` for the `get()` method?

I mean `onRemove` can be called by another thread as many thread can run the `get()` simultaneously.

linhong-intel replied:
if we add `synchronized` for `get()`, `onRemoval` can NOT be called by another thread. It's always called by `get()` and we have locked it.

chenghao-intel replied:
yes, I agreed with u, sorry for the confusion. :)

linhong-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -46,6 +46,39 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
+  private val DEFAULT_QUEUE_INITIAL_CAPACITY = 11
+  // Ordering.by() makes sure non-used FiberCache dequeue first, used by test suite.
+  private[filecache] val removalPendingQueue = new PriorityBlockingQueue[FiberCache](
+    DEFAULT_QUEUE_INITIAL_CAPACITY, Ordering.by(_.refCount))
```
you mean `ConcurrentLinkedQueue` ? yes, we can.
But then the `CacheGuardian` should change from signal wake up to periodically wake up.

chenghao-intel replied:
As you see below, we've added the wake up in the `FiberCache.tryDispose()`, so it's not a problem.

BTW, since the queue actually keep adding the blocks by LRU from the Guava, hence probably a normal `ConcurrentLinkedQueue` is good enough for now, and it supposedly that the Fiber in the head of the queue will be the longest time not been used.

linhong-intel replied:
sure.

linhong-intel commented:
@chenghao-intel , @carsonwang @yueawang  for review. I've addressed most reviews.
here is a summary:
1. record pending cache size -> done
2. Change to `synchronized FiberCache` for `occupy`, `release`, `dispose` - > done
3. `synchronized FiberCacheManager.get()` -> done
4. Use `ConcurrentLinkedQueue` -> done
5. release function for CacheGuardian and FiberCacheManager -> added a TODO
6. More test case -> WIP
7. OOM may happen in current PR -> Shall we fix this in release 0.3 or just leave a log?

chenghao-intel commented:
Thank you @linhong-intel for the summary, for 5), probably you have to change if you meet any problem in writing the unit tests, but I am OK if you decide to put it into next release. for 7) let's just leave a warning log in this release.

chenghao-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -46,6 +46,55 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
+
+    override def run(): Unit = {
+      // Loop forever, TODO: provide a release function
+      while (true) {
```
Sorry, probably `LinkedBlockingQueue` is much better:
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/LinkedBlockingQueue.html

It provide the 
Is the take API a block API? Or the code like:
```scala
while (true) {
  val cache = removalPendingQueue.poll(200, TimeUnit.MILLISECONDS)
  if (cache != null) {
    ...
  }
  ...
}
```

linhong-intel replied:
agree, I also thought about `LinkedBlockingQueue`

chenghao-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -46,6 +46,55 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
+      removalPendingQueue.add(fiberCache)
+    }
+
+    def removalPendingQueueSize: Int = removalPendingQueue.size()
```
A better name will be `pendingSize` for short?

linhong-intel replied:
sure

chenghao-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -46,6 +46,55 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
+            while (!fiberCache.tryDispose(3000)) {
+              // Check memory usage every 3s while we are waiting fiber release.
+              if (_pendingFiberSize.get() > maxMemory) {
+                logError("Fibers pending on removal use too much memory")
```
put this log into the `addRemovalFiber` function also?

chenghao-intel replied:
And it's better to print out the current and maximum memory in the log, which should be more informative to the end user.

linhong-intel replied:
sure

chenghao-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -46,6 +46,55 @@ class OapFiberCacheHeartBeatMessager extends CustomManager with Logging {
+        }
+        // Sleep 1s if there is no pending fibers
+        logDebug("Processed all pending removal fibers, sleep 1 seconds")
+        Thread.sleep(1000)
```
this is not necessary if we used the `BlockedQueue`

chenghao-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala:
```diff
@@ -92,17 +141,21 @@ object FiberCacheManager extends Logging {
+  private val cache = CacheBuilder
+    .newBuilder()
+    .recordStats()
+    .concurrencyLevel(4) // TODO: Make 4 configurable
```
I know this is from history reason, but is there any meaningful constant value name instead of `4` in Guava?

linhong-intel replied:
there is a `DEFAULT_CONCURRENCY_LEVEL` in guava, and it is 4. :(
Anyway, i'm considering to remove it since use `synchronized get` will lose concurrency

chenghao-intel commented on src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/MemoryManager.scala:
```diff
@@ -32,14 +33,44 @@ import org.apache.spark.unsafe.memory.{MemoryAllocator, MemoryBlock}
+    notifyAll()
+  }
+
+  def tryDispose(timeout: Long): Boolean = synchronized {
```
Nit: 
As most of `condition wait` will be a condition right before the `wait`, so I would like to suggest the code like:
```scala
val startTime = System.currentTimeMillis();
while (System.currentTimeMillis() - startTime <= timeout) {
  if (_refCount > 0) {
    try { wait(100) } catch {..}
  } else {
    realDispose()
    return true;
  }
}
return false;
```

chenghao-intel commented:
@linhong-intel I leave some more comments, maybe you can take some of the suggestions in the updated code. please at me when you have done. I didn't take too much code in code review, but it's worth to well designed.  :)

linhong-intel commented:
@chenghao-intel 
updated.
1 note:
I removed `.concurrencyLevel(4)`. for 2 reasons:
cache default concurrency level is 4 and use `synchronized get` will lose concurrency anyway

linhong-intel commented:
Also updated some test cases focused on race condition, like:
multiple threads get 1 fiber
multiple threads get multiple fiber
multiple threads release 1 fiber
multiple threads get/release fiber at the same time
multiple threads get/remove fiber at the same time

carsonwang commented on src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManagerSuite.scala:
```diff
@@ -47,41 +51,182 @@ class FiberCacheManagerSuite extends SharedOapContext {
+    val data = generateData(kbSize)
+    val fiber = TestFiber(() => MemoryManager.putToDataFiberCache(data), s"release test")
+    val fiberCaches = (1 to 5).map(_ => FiberCacheManager.get(fiber, configuration))
+    fiberCaches.foreach{ fiberCache =>
```
nit: need one space before {

carsonwang commented on src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManagerSuite.scala:
```diff
@@ -47,41 +51,182 @@ class FiberCacheManagerSuite extends SharedOapContext {
+    }
+    val threads = (0 until 5).map(i => new FiberTestRunner(i))
+    threads.foreach(_.start())
+    threads.foreach(_.join())
```
No assert needed in this test? Do we want to fail it after a timeout?

carsonwang commented on src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManagerSuite.scala:
```diff
@@ -47,41 +51,182 @@ class FiberCacheManagerSuite extends SharedOapContext {
+    }
+    val removeWork: Unit = {
+      (1 to 1000).foreach { _ =>
+        FiberCacheManager.removeFiber(fiber)
```
Do we need `removeFiber` exactly after `FiberCacheManager.get` and check the fiber is still not disposed after a few seconds? Can we do that in a single thread as `removeFiber` only invalidate the cache and move the fiber to the pending removal queue? It seems we can't guarantee when `removeFilber` is called if it is in another thread? I mean `removeFiber` may be called after our assert.

linhong-intel replied:
In out, "get" fiber from cache and "occupy" fiber are not atomic.
I just want to make sure these 2 steps are protected and no remove can happen between them.
I'll talk with you about this offline.

carsonwang commented on src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManagerSuite.scala:
```diff
@@ -47,41 +51,182 @@ class FiberCacheManagerSuite extends SharedOapContext {
+    }
+    pool.shutdown()
+    pool.awaitTermination(1000, TimeUnit.MILLISECONDS)
+    assert(fiberCaches.head.refCount == 0)
```
Do we need assert the refCount is 5 initially?

linhong-intel replied:
done
